cinn/frontend/pass/transpose_folding.cc

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2022 CINN Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cinn/frontend/cinn_builder.h"
+#include "cinn/frontend/program_pass.h"
+
+namespace cinn::frontend::pass {
+
+// Rules that transpose can fold into dot:
+//   1) input operand of dot must be transpose;
+//   2) `axis` of tranpose must be consecutive;
+std::vector<Instruction*> TryFoldIntoDot(Instruction* dot,
+                                         const absl::flat_hash_map<std::string, Instruction*>& var_instrs) {
+  std::vector<Instruction*> remove_instrs;
+  if (!(*dot)->attrs.empty()) return remove_instrs;
+  auto is_transpose = [](const Instruction& transpose) {
+    if ("transpose" != transpose->op_type) return false;
+    auto axis = transpose.GetAttrs<std::vector<int>>("axis");
+    for (size_t i = 0; i < axis.size(); ++i) {
+      if (axis[i] + 1 != axis.size() - i) return false;
+    }
+    return true;
+  };
+  for (size_t i = 0; i < (*dot)->inputs.size(); ++i) {
+    auto iter = var_instrs.find((*dot)->inputs[i]->id);
+    if (iter != var_instrs.end()) {
+      const auto& operand = *(iter->second);
+      if (is_transpose(operand)) {
+        // x-> transpose -> out -> dot => x -> dot
+        (*dot)->inputs[i] = operand->inputs[0];
+        (*dot).SetAttr(i == 0 ? "trans_a" : "trans_b", true);
+        remove_instrs.push_back(iter->second);
+      }
+    }
+  }
+  return remove_instrs;
+}
+
+// Pass `TransposeFolding` folds transpose into dot, than it can be implemented by a GEMM kernel.
```

Comment:
than --> then

then both of them can be ...

Reply:
done.